ksql-api/src/main/java/io/confluent/ksql/api/server/InsertsStreamHandler.java

```diff
@@ -103,30 +122,55 @@ private void handleArgs(final Buffer buff) {
         return;
       }
 
-      acksSubscriber =
-          insertsStreamArgs.get().requiresAcks ? new AcksSubscriber(ctx, routingContext.response(),
-              insertsStreamResponseWriter) : null;
-      final InsertsSubscriber insertsSubscriber = endpoints
-          .createInsertsSubscriber(insertsStreamArgs.get().target,
-              insertsStreamArgs.get().properties,
-              acksSubscriber);
-      publisher = new BufferedPublisher<>(ctx);
+      routingContext.response().endHandler(v -> handleResponseEnd());
+
+      acksSubscriber = new AcksSubscriber(ctx, routingContext.response(),
+          insertsStreamResponseWriter);
+
+      recordParser.pause();
+      createInsertsSubscriberAsync(insertsStreamArgs.get().target,
+          insertsStreamArgs.get().properties,
+          acksSubscriber, ctx)
+          .thenAccept(insertsSubscriber -> {
+            publisher = new BufferedPublisher<>(ctx);
 
-      // This forces response headers to be written so we know we send a 200 OK
-      // This is important if we subsequently find an error in the stream
-      routingContext.response().write("");
+            // This forces response headers to be written so we know we send a 200 OK
+            // This is important if we subsequently find an error in the stream
+            routingContext.response().write("");
 
-      publisher.subscribe(insertsSubscriber);
+            publisher.subscribe(insertsSubscriber);
 
+            recordParser.resume();
+
+            this.insertsSubscriber = insertsSubscriber;
+          })
+          .exceptionally(t -> handleInsertSubscriberException(t, routingContext));
     }
 
-    private void handleRow(final Buffer buff) {
+    private InsertsStreamSubscriber insertsSubscriber;
```

Comment:
nit: Can we have all the fields declared at the top of the file, please?

Reply:
ack